File: AppWidgetTest/app/src/main/java/com/example/appwidgettest/MainActivity.java
package com.example.appwidgettest;

import androidx.appcompat.app.AppCompatActivity;

import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "AppWidgetActivity";
    private static final String PREFS_NAME = "com.example.appwidgettest.AppWidgetTest";
    private static final String PREF_PREFIX_KEY = "prefix_";

    int mAppWidgetId = AppWidgetManager.INVALID_APPWIDGET_ID;
    EditText mAppWidgetPrefix;

    public MainActivity(){
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setResult(RESULT_CANCELED);
        setContentView(R.layout.activity_main);

        mAppWidgetPrefix = findViewById(R.id.appwidget_prefix);
        findViewById(R.id.save_button).setOnClickListener(mOnClickListener);

        Intent intent = getIntent();
        Bundle extras = intent.getExtras();
        if (extras != null){
            mAppWidgetId = extras.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
        }

        if (mAppWidgetId == AppWidgetManager.INVALID_APPWIDGET_ID){
            finish();
        }

        mAppWidgetPrefix.setText(loadTitlePref(MainActivity.this, mAppWidgetId));
    }

    View.OnClickListener mOnClickListener = new View.OnClickListener(){
        @Override
        public void onClick(View v){
            final Context context = MainActivity.this;

            String titlePrefix = mAppWidgetPrefix.getText().toString();
            //保存内容
            saveTitlePref(context, mAppWidgetId, titlePrefix);

            AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
            //更新微件
            ExampleAppWidgetProvider.updateAppWidget(context, appWidgetManager, mAppWidgetId, titlePrefix);

            Intent result = new Intent();
            result.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, mAppWidgetId);
            setResult(RESULT_OK, result);
            finish();
        }
    };

    private void saveTitlePref(Context context, int appWidgetId, String text){
        SharedPreferences.Editor sp = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).edit();
        sp.putString(PREF_PREFIX_KEY + appWidgetId, text);
        sp.commit();
    }

    public static String loadTitlePref(Context context, int appWidgetId){
        SharedPreferences sp = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        String text = sp.getString(PREF_PREFIX_KEY + appWidgetId, null);

        if (text != null) {
            return text;
        } else {
            return context.getString(R.string.appwidget_prefix_default);
        }
    }

    public static void loadAllTitlePrefs(Context context, ArrayList<Integer> appWidgetIds, ArrayList<String> texts) {

    }

    public static void deleteTitlePrefs(Context context, int appWidgetId){

    }
}

File: KotlinPratice/KotlinAndroid/app/src/main/java/com/sword/kotlinandroid/MainActivity.kt
package com.sword.kotlinandroid

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import com.sword.base.utils.CacheUtils
import com.sword.base.utils.toast
import com.sword.kotlinandroid.entity.User
import com.sword.kotlinandroid.lesson.LessonActivity
import com.sword.kotlinandroid.widget.CodeView


class MainActivity : AppCompatActivity(), View.OnClickListener {
    private val usernameKey: String = "username"
    private val passwordKey: String = "password"

    private lateinit var et_username: EditText
    private lateinit var et_password: EditText
    private lateinit var et_code: EditText


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        et_username = findViewById(R.id.et_username)
        et_password = findViewById(R.id.et_password)
        et_code = findViewById(R.id.et_code)

        et_username.setText(CacheUtils.get(usernameKey))
        et_password.setText(CacheUtils.get(passwordKey))

        val btn_login: Button = findViewById(R.id.btn_login)
        val img_code: CodeView = findViewById(R.id.code_view)

        btn_login.setOnClickListener(this)
        img_code.setOnClickListener(this)

    }

    @RequiresApi(Build.VERSION_CODES.M)
    override fun onClick(v: View?) {
        if (v is CodeView) {
            v.updateCode()
        } else if (v is Button) {
            login()
        }
    }

    private fun login() {
        val username: String = et_username.text.toString()
        val password: String = et_password.text.toString()
        val code: String = et_code.text.toString()

        val user = User(username, password, code)

        if (verify(user)) {
            Log.d("SWORD", "LOGIN VERIFY PASS")
            CacheUtils.save(usernameKey, username)
            CacheUtils.save(passwordKey, password)
            startActivity(Intent(this, LessonActivity::class.java))
        }

    }

    private fun verify(user: User): Boolean {
        if(user.username?.length?:0 < 4){
            toast("用户名不合法")
            return false
        }

        if (user.password?.length?:0 < 4) {
            toast("密码不合法")
            return false
        }
        return true
    }
}